src/mol-repr/volume/gradient.ts

```diff
@@ -0,0 +1,726 @@
+                    // add point (index coords; geometry stage does gridToCartn)
+                    const pw = Vec3();
+                    toIndexPoint(li, lj, lk, fx, fy, fz, pw);
+                    out[segIdx].push({ x: pw[0], y: pw[1], z: pw[2], value: level });
+                    nVert++;
+                }
+
+                // record visited cell for oblation (store once per change)
+                const n = activeCells.length;
+                if (n < 3 || activeCells[n-3] !== li || activeCells[n-2] !== lj || activeCells[n-1] !== lk) {
+                    activeCells.push(li, lj, lk);
+                }
+
+                // directional coherence (avoid flip-flop)
+                if (havePrev) {
+                    const dp = gW[0]*prev[0] + gW[1]*prev[1] + gW[2]*prev[2];
+                    if (dp < minDot) break;
+                }
+                Vec3.copy(prev, gW); havePrev = true;
+
+                // advance in INDEX coordinates using a world step
+                fx += gW[0] * (dsWorld / hx);
+                fy += gW[1] * (dsWorld / hy);
+                fz += gW[2] * (dsWorld / hz);
+
+                // decimate writes
+                if ((step % writeStride) !== 0) {
+                    // nothing; sampling throttled by outer segment build
+                }
+            }
+
+            // discard degenerate polylines
+            if (nVert < 2) {
+                out[segIdx] = [] as StreamlinePoint[]; // keep slot empty; harmless to geometry stage
+            }
+        };
+
+        walk(+1);
+        walk(-1);
+
+        // oblation (flag cells in a sphere of radius R around visited cells)
+        const R = OBLATION_SPACING|0; const R2 = R*R;
+        for (let t = 0; t < activeCells.length; t += 3) {
+            const ii = activeCells[t], jj = activeCells[t+1], kk = activeCells[t+2];
+            for (let k = Math.max(k0, kk-R); k <= Math.min(k1-1, kk+R); k++) {
+                const dz2 = (kk-k)*(kk-k);
+                for (let j = Math.max(j0, jj-R); j <= Math.min(j1-1, jj+R); j++) {
+                    const dy2 = (jj-j)*(jj-j) + dz2; if (dy2 > R2) continue;
+                    for (let i = Math.max(i0, ii-R); i <= Math.min(i1-1, ii+R); i++) {
+                        const d2 = (ii-i)*(ii-i) + dy2; if (d2 <= R2) flag[flagIndex(i,j,k)] = 1;
+                    }
+                }
+            }
+        }
+    }
+}
+
+// ---------------------------------------------------------------------------------------
+// Field sampling (clamped + world-aware gradients)
+// ---------------------------------------------------------------------------------------
+
+function clampIndex(space: Tensor.Space, x: number, y: number, z: number) {
+    const nx = space.dimensions[0] as number, ny = space.dimensions[1] as number, nz = space.dimensions[2] as number;
+    return [Math.max(0, Math.min(nx - 1, x)), Math.max(0, Math.min(ny - 1, y)), Math.max(0, Math.min(nz - 1, z))] as [number, number, number];
```

Comment:
`clampIndex` seems to be called inside a loop -- this should get an `out: [number, number, number]` vector instead as a param too.